web/src/dropdown_widget.js

```diff
@@ -1,176 +1,271 @@
+                        $element.trigger("focus");
+                    }
+
+                    const $search_input = $popper.find(".dropdown-list-search-input");
+                    const list_items = this.list_widget.get_current_list();
+                    if (list_items.length === 0 && !(e.key === "Escape")) {
+                        // Let the browser handle it.
+                        return;
+                    }
+
+                    function first_item() {
+                        const first_item = list_items[0];
+                        return $popper.find(`.list-item[data-unique-id="${first_item.unique_id}"]`);
+                    }
+
+                    function last_item() {
+                        const last_item = list_items.at(-1);
+                        return $popper.find(`.list-item[data-unique-id="${last_item.unique_id}"]`);
+                    }
+
+                    const render_all_items_and_focus_last_item = function () {
+                        // List widget doesn't render all items by default, so we need to render all
+                        // the items and focus on the last element.
+                        const list_items = this.list_widget.get_current_list();
+                        this.list_widget.render(list_items.length);
+                        trigger_element_focus(last_item());
+                    }.bind(this);
+
+                    switch (e.key) {
+                        case "Enter":
+                            if (e.target === $search_input.get(0)) {
+                                // Select first item if in search input.
+                                first_item().trigger("click");
+                            } else if (list_items.length !== 0) {
+                                $(e.target).trigger("click");
+                            }
+                            e.stopPropagation();
+                            e.preventDefault();
+                            break;
+
+                        case "Escape":
+                            instance.hide();
+                            this.on_exit_with_escape_callback();
+                            e.stopPropagation();
+                            e.preventDefault();
+                            break;
+
+                        case "Tab":
+                        case "ArrowDown":
+                            switch (e.target) {
+                                case last_item().get(0):
+                                    trigger_element_focus($search_input);
+                                    break;
+                                case $search_input.get(0):
+                                    trigger_element_focus(first_item());
+                                    break;
+                                default:
+                                    trigger_element_focus($(e.target).next());
+                            }
+                            break;
+
+                        case "ArrowUp":
```

Comment:
Do we have a means for handling Shift + Tab? Seems like if Tab moves down, Shift + Tab should move up.

Reply:
This is mostly focused around making arrow keys work. I can add functionality for it, but this is out of scope for this PR.